src/aliceVision/depthMap/cuda/commonStructures.hpp

```diff
@@ -452,23 +567,39 @@ template<class Type, unsigned Dim> void copy(CudaHostMemoryHeap<Type, Dim>& _dst
     for(unsigned i = 3; i < Dim; ++i)
       p.extent.depth *= _src.getSize()[i];
     p.kind = kind;
-    cudaMemcpy3D(&p);
+    cudaError_t err = cudaMemcpy3D(&p);
+    THROW_ON_CUDA_ERROR(err, "Failed to copy (" << __FILE__ << " " << __LINE__ << ")");
   }
 }
 
 template<class Type, unsigned Dim> void copy(CudaDeviceMemoryPitched<Type, Dim>& _dst, const CudaHostMemoryHeap<Type, Dim>& _src)
 {
   cudaMemcpyKind kind = cudaMemcpyHostToDevice;
   if(Dim == 1) {
-    cudaMemcpy(_dst.getBuffer(), _src.getBuffer(), _src.getBytes(), kind);
+    cudaError_t err = cudaMemcpy(_dst.getBuffer(), _src.getBuffer(), _src.getBytes(), kind);
+    THROW_ON_CUDA_ERROR(err, "Failed to copy (" << __FILE__ << " " << __LINE__ << ")");
   }
   else if(Dim == 2) {
-    cudaMemcpy2D(_dst.getBuffer(), _dst.getPitch(), _src.getBuffer(), _src.getSize()[0] * sizeof (Type), _src.getSize()[0] * sizeof (Type), _src.getSize()[1], kind);
+    cudaError_t err = cudaMemcpy2D(_dst.getBuffer(),
+                                   _dst.getPitch(),
+                                   _src.getBuffer(),
+                                   _src.getSize()[0] * sizeof (Type),
+                                   _src.getSize()[0] * sizeof (Type),
+                                   _src.getSize()[1],
+                                   kind);
+    THROW_ON_CUDA_ERROR(err, "Failed to copy (" << __FILE__ << " " << __LINE__ << ")");
   }
   else if(Dim >= 3) {
     for (unsigned int slice=0; slice<_src.getSize()[2]; slice++)
     {
-      cudaMemcpy2D( &_dst.getBuffer()[slice * _dst.stride()[1]], _dst.getPitch(), _src.getBuffer() + slice * _src.getSize()[0] * _src.getSize()[1], _src.getSize()[0] * sizeof (Type), _src.getSize()[0] * sizeof (Type), _src.getSize()[1], kind);
+      cudaError_t err = cudaMemcpy2D( &_dst.getBuffer()[slice * _dst.stride()[1]],
+                                      _dst.getPitch(),
+                                      _src.getBuffer() + slice * _src.getSize()[0] * _src.getSize()[1],
```

Comment:
same question here

Reply:
Well ... I see why this looks weird.

stride() returns a Dim-dimensional array of (width, height, depth) in bytes. getSize() returns (width, height, depth) in number of elements. Of course, numbers are the same for all dimensions except dim 0, width.

On the device side, we must use Stride because of the padding that CUDA adds to "width" (X dimension) during allocation. On the host side, we could also use stride - except that the data structure on the host side doesn't have the stride() function so far. I could add it, of course, gaining a bit of readability but spending a useless multiplication.

We could re-interpret stride() on the GPU side, by dividing the number by sizeof(Type). However, nothing guarantees us that the stride can be divided by sizeof(Type). It probably does. But we cannot be sure.

Reply:
We could use explicit naming, like `strideBytes()` and `stride()` if it's difficult to use the same unit in all classes.
What do you think?